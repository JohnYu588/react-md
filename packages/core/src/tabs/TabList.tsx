"use client";
import { cnb } from "cnbuilder";
import type { HTMLAttributes } from "react";
import { forwardRef, useEffect, useRef, useState } from "react";
import { useAppSize } from "../AppSizeProvider.js";
import { KeyboardMovementProvider } from "../movement/useKeyboardMovementProvider.js";
import type { UseStateSetter } from "../types.js";
import { bem } from "../utils/bem.js";
import type { BaseTabListScrollButtonProps } from "./TabListScrollButton.js";
import { TabListScrollButton } from "./TabListScrollButton.js";
import { useTabList } from "./useTabList.js";
import type { GetTabListScrollToOptions } from "./utils.js";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { useTabs } from "./useTabs.js";

const styles = bem("rmd-tablist");

/**
 * The supported types of alignments for the tabs.
 */
export type TabsAlignment = "left" | "center" | "right";

/**
 * @remarks \@since 6.0.0
 */
export interface TabListClassNameOptions {
  className?: string;
  align?: TabsAlignment;
  animate?: boolean;
  padded?: boolean;
  vertical?: boolean;
  scrollbar?: boolean;
}

/**
 * @remarks \@since 6.0.0
 */
export function tabList(options: TabListClassNameOptions = {}): string {
  const {
    className,
    align = "left",
    padded,
    vertical,
    scrollbar,
    animate = false,
  } = options;

  return cnb(
    styles({
      [align]: true,
      padded,
      vertical,
      animate,
      "no-scrollbar": !scrollbar,
    }),
    className
  );
}

/**
 * @remarks \@since 6.0.0
 */
export interface TabListProps extends HTMLAttributes<HTMLDivElement> {
  activeIndex: number;
  setActiveIndex: UseStateSetter<number>;

  /**
   * @defaultValue `"left"`
   */
  align?: TabsAlignment;

  /**
   * @defaultValue `false`
   */
  padded?: boolean;

  /**
   * @defaultValue `false`
   */
  vertical?: boolean;

  /**
   * Set this to `true` to show a scrollbar when the number of tabs cause
   * overflow.
   *
   * @see {@link scrollButtons}
   * @defaultValue `false`
   */
  scrollbar?: boolean;

  /**
   * Set this to `true` to disable the active tab indicator from animating
   * when a new tab is selected.
   *
   * @defaultValue `false`
   */
  disableTransition?: boolean;

  /**
   * This should be equal to the `$tabs-transition-duration` variable.
   *
   * @defaultValue `150`
   */
  transitionDuration?: number;

  /**
   * @defaultValue `"manual"`
   */
  activationMode?: "manual" | "automatic";

  /**
   * Set this to `true` to render buttons that can scroll forwards or backwards
   * within the tab list if there is overflow **on desktop**. If you want to
   * display the scroll buttons on mobile as well, set this to `"allow-phone"` .
   *
   * @defaultValue `false`
   */
  scrollButtons?: boolean | "allow-phone";

  /**
   * A convenience prop for the {@link BaseTabListScrollButtonProps.getScrollToOptions}
   * on {@link forwardScrollButtonProps} and {@link backwardScrollButtonProps}.
   */
  getScrollToOptions?: GetTabListScrollToOptions;

  /**
   * Any additional props that should be passed to the scroll forward button.
   *
   * @example
   * ```tsx
   * forwardScrollButtonProps={{
   *   "aria-label": "Scroll right",
   *   theme: "primary",
   *   themeType: "contained",
   *   className: styles.buttonContainer,
   *   buttonProps: {
   *     className: styles.button,
   *   }
   * }}
   * ```
   */
  forwardScrollButtonProps?: BaseTabListScrollButtonProps;

  /**
   * Any additional props that should be passed to the scroll backward button.
   *
   * @example
   * ```tsx
   * forwardScrollButtonProps={{
   *   "aria-label": "Scroll left",
   *   theme: "primary",
   *   themeType: "contained",
   *   className: styles.buttonContainer,
   *   buttonProps: {
   *     className: styles.button,
   *   }
   * }}
   * ```
   */
  backwardScrollButtonProps?: BaseTabListScrollButtonProps;
}

/**
 * **Client Component**
 *
 * @see {@link useTabs} for example usage.
 *
 * @remarks \@since 6.0.0
 */
export const TabList = forwardRef<HTMLDivElement, TabListProps>(
  function TabList(props, ref) {
    const {
      style,
      onClick,
      onFocus,
      onKeyDown,
      className,
      children,
      activeIndex,
      setActiveIndex,
      activationMode = "manual",
      align = "left",
      padded = false,
      vertical = false,
      scrollbar = false,
      scrollButtons = false,
      disableTransition = false,
      transitionDuration = 150,
      getScrollToOptions,
      forwardScrollButtonProps,
      backwardScrollButtonProps,
      ...remaining
    } = props;

    const { isPhone } = useAppSize();
    const showScrollButtons =
      !!scrollButtons && (scrollButtons === "allow-phone" || !isPhone);

    const { elementProps, movementContext, backwardProps, forwardProps } =
      useTabList({
        ref,
        style,
        onClick,
        onFocus,
        onKeyDown,
        vertical,
        activeIndex,
        setActiveIndex,
        activationMode,
        scrollButtons: showScrollButtons,
      });

    const prevActiveIndex = useRef(activeIndex);
    const [animate, setAnimate] = useState(false);
    useEffect(() => {
      const isSameIndex = activeIndex === prevActiveIndex.current;
      prevActiveIndex.current = activeIndex;
      if (disableTransition || isSameIndex) {
        return;
      }

      setAnimate(true);
      const timeout = window.setTimeout(() => {
        setAnimate(false);
      }, transitionDuration);

      return () => {
        window.clearTimeout(timeout);
      };
    }, [activeIndex, disableTransition, transitionDuration]);

    return (
      <KeyboardMovementProvider value={movementContext}>
        <div
          {...remaining}
          {...elementProps}
          role="tablist"
          className={tabList({
            align,
            padded,
            animate: !disableTransition && animate,
            vertical,
            scrollbar,
            className,
          })}
        >
          {showScrollButtons && (
            <TabListScrollButton
              getScrollToOptions={getScrollToOptions}
              {...backwardScrollButtonProps}
              {...backwardProps}
            />
          )}
          {children}
          {showScrollButtons && (
            <TabListScrollButton
              getScrollToOptions={getScrollToOptions}
              {...forwardScrollButtonProps}
              {...forwardProps}
            />
          )}
        </div>
      </KeyboardMovementProvider>
    );
  }
);
