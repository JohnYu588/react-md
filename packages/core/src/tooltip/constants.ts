import {
  type CSSTransitionClassNames,
  type TransitionTimeout,
} from "../transition/types.js";

/** @since 2.0.0 */
export const DEFAULT_TOOLTIP_TIMEOUT = {
  enter: 200,
  exit: 150,
} as const satisfies TransitionTimeout;

/** @since 2.0.0 */
export const DEFAULT_TOOLTIP_CLASSNAMES = {
  appear: "rmd-tooltip--enter",
  appearActive: "rmd-tooltip--visible",
  enter: "rmd-tooltip--enter",
  enterActive: "rmd-tooltip--visible",
  enterDone: "rmd-tooltip--visible",
  exit: "rmd-tooltip--visible rmd-tooltip--exit",
  exitActive: "rmd-tooltip--exit-active",
} as const satisfies CSSTransitionClassNames;

/** @since 2.0.0 */
export const DEFAULT_TOOLTIP_DELAY = 1000;
/** @since 2.0.0 */
export const DEFAULT_TOOLTIP_THRESHOLD = 0.75;

/** @since 2.8.0 */
export const DEFAULT_TOOLTIP_SPACING = "1.5rem";
/** @since 2.8.0 */
export const DEFAULT_TOOLTIP_DENSE_SPACING = "0.875rem";
/** @since 2.8.0 */
export const DEFAULT_TOOLTIP_MARGIN = 16;
/** @since 2.8.0 */
export const DEFAULT_TOOLTIP_POSITION = "below";

/**
 * @internal
 * @since 2.8.0
 */
export const TOOLTIP_SPACING_VAR = "--rmd-tooltip-spacing";
