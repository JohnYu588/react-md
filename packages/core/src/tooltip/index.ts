export * from "./constants";
export * from "./Tooltip";
export * from "./TooltipHoverModeProvider";
export * from "./useTooltip";
