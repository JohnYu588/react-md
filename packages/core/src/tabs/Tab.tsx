"use client";
import {
  type AnchorHTMLAttributes,
  type ButtonHTMLAttributes,
  type ReactElement,
  type ReactNode,
} from "react";
import { type ComponentWithRippleProps } from "../interaction/types.js";
import { useElementInteraction } from "../interaction/useElementInteraction.js";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren.js";
import { type CustomLinkComponent } from "../link/Link.js";
import { useKeyboardMovementContext } from "../movement/useKeyboardMovementProvider.js";
import { useEnsuredId } from "../useEnsuredId.js";
import { tab } from "./tabStyles.js";

// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { type useTabs } from "./useTabs.js";
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { type TabListProps } from "./TabList.js";

/**
 * @since 6.0.0
 */
export interface BaseTabProps extends ComponentWithRippleProps {
  /**
   * Set this to `true` if the tab is currently active.
   *
   * This is normally provided by the {@link useTabs} hook.
   */
  active: boolean;

  /**
   * Set this to `true` if the {@link TabListProps.disableTransition} prop has
   * also been set to `true` to disable an active indicator below the tab when
   * {@link active} is `true`.
   *
   * @defaultValue `false`
   */
  activeIndicator?: boolean;

  /**
   * Set this to `true` when rendering the tabs vertically and
   * {@link activeIndicator} has been enabled.
   *
   * @defaultValue !false
   */
  verticalActiveIndicator?: boolean;

  /**
   * An optional icon to render with the with the {@link children}. The default
   * behavior will render this icon before the children.
   *
   * @see {@link iconAfter}
   * @see {@link stacked}
   */
  icon?: ReactNode;

  /**
   * Set this to `true` to render the {@link icon} after the {@link children}.
   *
   * @defaultValue `false`
   */
  iconAfter?: boolean;

  /**
   * Set this to `true` to render the {@link icon} and {@link children} stacked
   * instead of horizontally.
   *
   * @defaultValue `false`
   */
  stacked?: boolean;
}

/**
 * @since 6.0.0
 */
export interface TabButtonProps
  extends BaseTabProps,
    ButtonHTMLAttributes<HTMLButtonElement> {
  as?: "button";
}

/**
 * @since 6.0.0
 */
export interface TabLinkProps
  extends BaseTabProps,
    AnchorHTMLAttributes<HTMLAnchorElement> {
  as: CustomLinkComponent;
}

/**
 * @since 6.0.0
 */
export type TabProps = TabButtonProps | TabLinkProps;

/**
 * **Client Component**
 *
 * This component should usually be used with the `TabsList` component and
 * `useTabs` hook.
 *
 * @see {@link useTabs}
 *
 * @since 6.0.0
 */
export function Tab(props: TabProps): ReactElement {
  const {
    id: propId,
    as: Component = "button",
    active,
    activeIndicator,
    verticalActiveIndicator,
    icon,
    iconAfter,
    stacked,
    className,
    children: propChildren,
    onBlur,
    onClick,
    onKeyDown,
    onKeyUp,
    onMouseDown,
    onMouseUp,
    onMouseLeave,
    onDragStart,
    onTouchStart,
    onTouchEnd,
    onTouchMove,
    disableRipple,
    ...remaining
  } = props as TabButtonProps;
  const { disabled } = props as TabButtonProps;

  const id = useEnsuredId(propId, "tab");
  const { activeDescendantId } = useKeyboardMovementContext();
  const { ripples, handlers } = useElementInteraction({
    mode: disableRipple ? "none" : undefined,
    onBlur,
    onClick,
    onKeyDown,
    onKeyUp,
    onMouseDown,
    onMouseUp,
    onMouseLeave,
    onDragStart,
    onTouchStart,
    onTouchEnd,
    onTouchMove,
    disabled,
  });

  const isLink = Component !== "button";
  const children = useHigherContrastChildren(propChildren);
  let buttonOnlyProps: Record<string, unknown> | undefined;
  if (!isLink) {
    buttonOnlyProps = { type: "button" };
  }

  return (
    <Component
      {...remaining}
      {...buttonOnlyProps}
      {...handlers}
      aria-selected={active}
      id={id}
      role="tab"
      tabIndex={id === activeDescendantId ? 0 : -1}
      className={tab({
        className,
        active,
        isLink,
        stacked: !!icon && stacked,
        disabled,
        reversed: !!icon && iconAfter,
        activeIndicator,
        verticalActiveIndicator,
      })}
    >
      {icon}
      {children}
      {ripples}
    </Component>
  );
}
