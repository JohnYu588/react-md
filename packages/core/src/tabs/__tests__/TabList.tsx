import {
  act,
  fireEvent,
  render as baseRender,
  screen,
  waitFor,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import type { ReactElement } from "react";

import {
  DEFAULT_DESKTOP_LARGE_MIN_WIDTH,
  DEFAULT_DESKTOP_MIN_WIDTH,
  DEFAULT_PHONE_MAX_WIDTH,
} from "../../AppSizeProvider";
import { CoreProviders } from "../../CoreProviders";
import { Tooltip, useTooltip } from "../../tooltip";
import { Tab } from "../Tab";
import type { TabListProps } from "../TabList";
import { TabList } from "../TabList";
import type { ProvidedTabListProps } from "../useTabs";
import { useTabs } from "../useTabs";
import type { TabListScrollToOptions } from "../utils";

const render = (ui: ReactElement, rtl = false): ReturnType<typeof baseRender> =>
  baseRender(ui, {
    wrapper: ({ children }) => (
      <CoreProviders
        elementInteractionMode="none"
        defaultDir={rtl ? "rtl" : "ltr"}
      >
        {children}
      </CoreProviders>
    ),
  });

// make it so the back button defaults to intersecting
const getIntersectionRatio = jest.fn((target: Element): number =>
  target.nextElementSibling ? 1 : 0
);

function Test(
  props: Omit<TabListProps, keyof ProvidedTabListProps>
): ReactElement {
  const { getTabProps, getTabListProps } = useTabs();

  return (
    <TabList {...props} {...getTabListProps()}>
      <Tab {...getTabProps(0)}>Tab 1</Tab>
      <Tab {...getTabProps(1)}>Tab 2</Tab>
      <Tab {...getTabProps(2)}>Tab 3</Tab>
    </TabList>
  );
}

class MockedObserver implements globalThis.IntersectionObserver {
  root: Document | Element | null;
  rootMargin: string;
  thresholds: readonly number[];

  elements: Set<Element>;

  constructor(
    public callback: IntersectionObserverCallback,
    options: IntersectionObserverInit = {}
  ) {
    this.root = options.root || null;
    this.rootMargin = options.rootMargin || "";
    this.thresholds =
      typeof options.threshold === "number"
        ? [options.threshold]
        : options.threshold ?? [];

    this.elements = new Set();
  }

  observe(target: Element): void {
    this.elements.add(target);

    this.trigger();
  }

  unobserve(target: Element): void {
    this.elements.delete(target);
  }

  takeRecords = (): IntersectionObserverEntry[] => {
    const rect = document.body.getBoundingClientRect();
    return [...this.elements].map<IntersectionObserverEntry>((target) => ({
      time: Date.now(),
      target,
      boundingClientRect: rect,
      intersectionRatio: getIntersectionRatio(target),
      intersectionRect: rect,
      isIntersecting: false,
      rootBounds: null,
    }));
  };

  disconnect(): void {
    this.elements.clear();
  }

  trigger = (): void => {
    this.callback(this.takeRecords(), this);
  };
}

type Observer = (IntersectionObserver & { trigger(): void }) | undefined;

describe("TabList", () => {
  beforeEach(() => {
    getIntersectionRatio.mockImplementation((element) =>
      element.nextElementSibling ? 1 : 0
    );
    jest
      .spyOn(window, "IntersectionObserver")
      .mockImplementation(
        (callback, options) => new MockedObserver(callback, options)
      );
  });

  it("should update the tab indicator styles based on the active index", async () => {
    const offsetLeft = jest
      .spyOn(HTMLButtonElement.prototype, "offsetLeft", "get")
      .mockReturnValue(0);
    const offsetWidth = jest
      .spyOn(HTMLButtonElement.prototype, "offsetWidth", "get")
      .mockReturnValue(120);

    const user = userEvent.setup();
    const { getByRole } = render(<Test />);
    const tablist = getByRole("tablist");
    const tab1 = getByRole("tab", { name: "Tab 1" });
    const tab2 = getByRole("tab", { name: "Tab 2" });

    // it should always start to 33.33% and then calculate the correct width
    // based on the number of tabs
    expect(tablist.style.getPropertyValue("--rmd-tab-width")).toBe(
      `${100 / 3}%`
    );
    expect(tablist.style.getPropertyValue("--rmd-tab-offset")).toBe("0px");

    // need to use waitFor since ResizeObserver uses requestAnimationFrame
    await waitFor(() => {
      expect(tablist.style.getPropertyValue("--rmd-tab-width")).toBe("120px");
      expect(tablist.style.getPropertyValue("--rmd-tab-offset")).toBe("0px");
    });

    expect(tablist).toMatchSnapshot();

    offsetLeft.mockReturnValue(120);
    offsetWidth.mockReturnValue(90);

    await user.click(tab2);
    await waitFor(() => {
      expect(tablist.style.getPropertyValue("--rmd-tab-width")).toBe("90px");
      expect(tablist.style.getPropertyValue("--rmd-tab-offset")).toBe("120px");
    });

    offsetLeft.mockReturnValue(0);
    offsetWidth.mockReturnValue(120);

    await user.click(tab1);
    await waitFor(() => {
      expect(tablist.style.getPropertyValue("--rmd-tab-width")).toBe("90px");
      expect(tablist.style.getPropertyValue("--rmd-tab-offset")).toBe("120px");
    });
  });

  it("should support rendering scroll buttons to scroll horizontally by setting the scrollButtons prop to true", async () => {
    let backObserver: Observer;
    let forwardObserver: Observer;
    jest
      .spyOn(window, "IntersectionObserver")
      .mockImplementation((callback, options) => {
        if (!backObserver) {
          backObserver = new MockedObserver(callback, options);
          return backObserver;
        }

        if (!forwardObserver) {
          forwardObserver = new MockedObserver(callback, options);
          return forwardObserver;
        }

        throw new Error();
      });

    const user = userEvent.setup();
    const { getByRole } = render(<Test scrollButtons />);
    const tablist = getByRole("tablist");
    await waitFor(() => {
      expect(backObserver).toBeDefined();
      expect(forwardObserver).toBeDefined();
    });

    if (!backObserver || !forwardObserver) {
      throw new Error();
    }

    const scrollLeft = jest
      .spyOn(tablist, "scrollLeft", "get")
      .mockReturnValue(0);
    jest.spyOn(tablist, "scrollWidth", "get").mockReturnValue(100);

    const back = getByRole("button", { name: "back" });
    const forward = getByRole("button", { name: "forward" });

    expect(back).toBeDisabled();
    expect(forward).not.toBeDisabled();

    tablist.scrollTo ??= jest.fn();
    const scrollTo = jest.spyOn(tablist, "scrollTo").mockImplementation(() => {
      act(() => {
        backObserver?.trigger();
        forwardObserver?.trigger();
      });
    });
    getIntersectionRatio.mockImplementation(() => 0);

    await user.click(forward);

    expect(scrollTo).toHaveBeenCalledWith({
      left: 10,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();

    scrollLeft.mockReturnValue(10);
    await user.click(forward);
    expect(scrollTo).toHaveBeenCalledWith({
      left: 20,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();

    scrollLeft.mockReturnValue(20);
    getIntersectionRatio.mockImplementation((target) =>
      target.nextElementSibling ? 0 : 1
    );
    await user.click(forward);
    expect(scrollTo).toHaveBeenCalledWith({
      left: 30,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).toBeDisabled();

    scrollLeft.mockReturnValue(30);
    getIntersectionRatio.mockReturnValue(0);

    await user.click(back);
    expect(scrollTo).toHaveBeenCalledWith({
      left: 20,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();
  });

  it("should support rendering scroll buttons to scroll vertically by setting the scrollButtons prop to true", async () => {
    let backObserver: Observer;
    let forwardObserver: Observer;
    jest
      .spyOn(window, "IntersectionObserver")
      .mockImplementation((callback, options) => {
        if (!backObserver) {
          backObserver = new MockedObserver(callback, options);
          return backObserver;
        }

        if (!forwardObserver) {
          forwardObserver = new MockedObserver(callback, options);
          return forwardObserver;
        }

        throw new Error();
      });

    const user = userEvent.setup();
    const { getByRole } = render(<Test scrollButtons vertical />);
    const tablist = getByRole("tablist");
    await waitFor(() => {
      expect(backObserver).toBeDefined();
      expect(forwardObserver).toBeDefined();
    });

    if (!backObserver || !forwardObserver) {
      throw new Error();
    }

    const scrollTop = jest
      .spyOn(tablist, "scrollTop", "get")
      .mockReturnValue(0);
    jest.spyOn(tablist, "scrollHeight", "get").mockReturnValue(100);
    jest.spyOn(tablist, "scrollWidth", "get").mockReturnValue(40);

    const back = getByRole("button", { name: "back" });
    const forward = getByRole("button", { name: "forward" });

    expect(back).toBeDisabled();
    expect(forward).not.toBeDisabled();

    tablist.scrollTo ??= jest.fn();
    const scrollTo = jest.spyOn(tablist, "scrollTo").mockImplementation(() => {
      act(() => {
        backObserver?.trigger();
        forwardObserver?.trigger();
      });
    });
    getIntersectionRatio.mockImplementation(() => 0);

    await user.click(forward);

    expect(scrollTo).toHaveBeenCalledWith({
      top: 10,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();

    scrollTop.mockReturnValue(10);
    await user.click(forward);
    expect(scrollTo).toHaveBeenCalledWith({
      top: 20,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();

    scrollTop.mockReturnValue(20);
    getIntersectionRatio.mockImplementation((target) =>
      target.nextElementSibling ? 0 : 1
    );
    await user.click(forward);
    expect(scrollTo).toHaveBeenCalledWith({
      top: 30,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).toBeDisabled();

    scrollTop.mockReturnValue(30);
    getIntersectionRatio.mockReturnValue(0);

    await user.click(back);
    expect(scrollTo).toHaveBeenCalledWith({
      top: 20,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();
  });

  it("should support custom labels for the scroll buttons", () => {
    const { getByRole } = render(
      <Test
        scrollButtons
        backwardScrollButtonProps={{
          "aria-label": "Scroll left",
        }}
        forwardScrollButtonProps={{
          "aria-label": "Scroll right",
        }}
      />
    );

    const back = getByRole("button", { name: "Scroll left" });
    const forward = getByRole("button", { name: "Scroll right" });

    expect(back).toBeInTheDocument();
    expect(forward).toBeInTheDocument();
  });

  it('should support rendering the scrollButtons on mobile by setting scrollButtons to "allow-phone"', () => {
    type Listener = (event: MediaQueryListEvent) => void;
    const noop = (): void => {
      // do nothing
    };
    const handlers = new Map<string, Listener>();
    jest.spyOn(window, "matchMedia").mockImplementation((query) => ({
      media: query,
      matches:
        query.includes(`${DEFAULT_DESKTOP_MIN_WIDTH}`) ||
        query.includes(`${DEFAULT_DESKTOP_LARGE_MIN_WIDTH}`),
      onchange: noop,
      addListener: noop,
      removeListener: noop,
      removeEventListener: noop,
      dispatchEvent: () => false,
      addEventListener(_type: string, listener: EventListener) {
        handlers.set(query, listener);
      },
    }));

    const { getByRole, rerender } = render(<Test scrollButtons />);

    const back = getByRole("button", { name: "back" });
    const forward = getByRole("button", { name: "forward" });
    expect(back).toBeInTheDocument();
    expect(forward).toBeInTheDocument();

    act(() => {
      [...handlers.entries()].forEach(([query, listener]) => {
        listener({
          ...new Event("change"),
          matches: query.includes(`${DEFAULT_PHONE_MAX_WIDTH}`),
          media: query,
        });
      });
    });

    expect(back).not.toBeInTheDocument();
    expect(forward).not.toBeInTheDocument();

    rerender(<Test scrollButtons="allow-phone" />);
    expect(getByRole("button", { name: "back" })).toBeInTheDocument();
    expect(getByRole("button", { name: "forward" })).toBeInTheDocument();
  });

  it("should support scrolling correctly in RTL mode", async () => {
    let backObserver: Observer;
    let forwardObserver: Observer;
    jest
      .spyOn(window, "IntersectionObserver")
      .mockImplementation((callback, options) => {
        if (!backObserver) {
          backObserver = new MockedObserver(callback, options);
          return backObserver;
        }

        if (!forwardObserver) {
          forwardObserver = new MockedObserver(callback, options);
          return forwardObserver;
        }

        throw new Error();
      });

    const user = userEvent.setup();
    const { getByRole } = render(<Test scrollButtons />, true);
    const tablist = getByRole("tablist");
    await waitFor(() => {
      expect(backObserver).toBeDefined();
      expect(forwardObserver).toBeDefined();
    });

    if (!backObserver || !forwardObserver) {
      throw new Error();
    }

    const scrollLeft = jest
      .spyOn(tablist, "scrollLeft", "get")
      .mockReturnValue(0);
    jest.spyOn(tablist, "scrollWidth", "get").mockReturnValue(100);

    const back = getByRole("button", { name: "back" });
    const forward = getByRole("button", { name: "forward" });

    expect(back).toBeDisabled();
    expect(forward).not.toBeDisabled();

    tablist.scrollTo ??= jest.fn();
    const scrollTo = jest.spyOn(tablist, "scrollTo").mockImplementation(() => {
      act(() => {
        backObserver?.trigger();
        forwardObserver?.trigger();
      });
    });
    getIntersectionRatio.mockImplementation(() => 0);

    await user.click(forward);

    expect(scrollTo).toHaveBeenCalledWith({
      left: -10,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();

    scrollLeft.mockReturnValue(-10);
    await user.click(forward);
    expect(scrollTo).toHaveBeenCalledWith({
      left: -20,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();

    scrollLeft.mockReturnValue(-20);
    getIntersectionRatio.mockImplementation((target) =>
      target.nextElementSibling ? 0 : 1
    );
    await user.click(forward);
    expect(scrollTo).toHaveBeenCalledWith({
      left: -30,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).toBeDisabled();

    scrollLeft.mockReturnValue(-30);
    getIntersectionRatio.mockReturnValue(0);

    await user.click(back);
    expect(scrollTo).toHaveBeenCalledWith({
      left: -20,
      behavior: "smooth",
    });
    expect(back).not.toBeDisabled();
    expect(forward).not.toBeDisabled();
  });

  it("should support scrolling a custom distance with the getScrollToOptions prop", async () => {
    // just allow both buttons to always be enabled for this test
    getIntersectionRatio.mockReturnValue(0.5);
    const getScrollToOptions = jest.fn<
      ScrollToOptions | undefined,
      [TabListScrollToOptions]
    >(() => ({
      top: 10,
      left: 200,
      behavior: "auto",
    }));
    const user = userEvent.setup();
    const { getByRole } = render(
      <Test scrollButtons getScrollToOptions={getScrollToOptions} />
    );
    const tablist = getByRole("tablist");
    tablist.scrollTo ??= jest.fn();
    const scrollTo = jest.spyOn(tablist, "scrollTo");

    const back = getByRole("button", { name: "back" });
    const forward = getByRole("button", { name: "forward" });

    await user.click(forward);
    expect(getScrollToOptions).toHaveBeenCalledWith({
      isRTL: false,
      vertical: false,
      animate: true,
      container: tablist,
      increment: true,
    });
    expect(scrollTo).toHaveBeenCalledWith({
      top: 10,
      left: 200,
      behavior: "auto",
    });

    getScrollToOptions.mockReturnValue({
      top: 20,
      left: -110,
      behavior: "auto",
    });
    await user.click(back);
    expect(getScrollToOptions).toHaveBeenCalledWith({
      isRTL: false,
      vertical: false,
      animate: true,
      container: tablist,
      increment: false,
    });
    expect(scrollTo).toHaveBeenCalledWith({
      top: 20,
      left: -110,
      behavior: "auto",
    });
  });

  it("should support rendering scrollbars if the scrollbar prop is enabled", () => {
    const { getByRole, rerender } = render(<Test />);
    const tablist = getByRole("tablist");
    expect(tablist).toHaveClass("rmd-tablist--no-scrollbar");

    rerender(<Test scrollbar />);
    expect(tablist).not.toHaveClass("rmd-tablist--no-scrollbar");
  });

  it("should support adding tooltips to the scroll buttons", async () => {
    // allow both tto be enabled for this test
    getIntersectionRatio.mockReturnValue(0.5);
    function TooltipTest(): ReactElement {
      const forwardTooltip = useTooltip({
        hoverTime: 0,
      });
      const backwardTooltip = useTooltip({
        hoverTime: 0,
      });
      return (
        <>
          <Test
            scrollButtons
            forwardScrollButtonProps={{
              buttonProps: forwardTooltip.elementProps,
            }}
            backwardScrollButtonProps={{
              buttonProps: backwardTooltip.elementProps,
            }}
          />
          <Tooltip {...forwardTooltip.tooltipProps}>Scroll right</Tooltip>
          <Tooltip {...backwardTooltip.tooltipProps}>Scroll left</Tooltip>
        </>
      );
    }

    const { getByRole } = render(<TooltipTest />);
    const back = getByRole("button", { name: "back" });
    const forward = getByRole("button", { name: "forward" });

    expect(() => getByRole("tooltip")).toThrow();

    fireEvent.mouseEnter(back);
    await waitFor(() => {
      expect(
        screen.getByRole("tooltip", { name: "Scroll left" })
      ).toBeInTheDocument();
    });

    fireEvent.mouseLeave(back);
    await waitFor(() => {
      expect(() => getByRole("tooltip")).toThrow();
    });

    fireEvent.mouseEnter(forward);
    await waitFor(() => {
      expect(
        screen.getByRole("tooltip", { name: "Scroll right" })
      ).toBeInTheDocument();
    });

    fireEvent.mouseLeave(forward);
    await waitFor(() => {
      expect(() => getByRole("tooltip")).toThrow();
    });
  });
});
