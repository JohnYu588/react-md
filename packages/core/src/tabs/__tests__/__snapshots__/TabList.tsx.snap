// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabList should update the tab indicator styles based on the active index 1`] = `
<div
  aria-orientation="horizontal"
  class="rmd-tablist rmd-tablist--left rmd-tablist--no-scrollbar"
  role="tablist"
  style="--rmd-tab-width: 120px; --rmd-tab-offset: 0px;"
  tabindex="0"
>
  <button
    aria-controls="tab-:r0:-panel-1"
    aria-selected="true"
    class="rmd-tab rmd-tab--active"
    id="tab-:r0:-1"
    role="tab"
    tabindex="-1"
    type="button"
  >
    <span>
      Tab 1
    </span>
  </button>
  <button
    aria-controls="tab-:r0:-panel-2"
    aria-selected="false"
    class="rmd-tab"
    id="tab-:r0:-2"
    role="tab"
    tabindex="-1"
    type="button"
  >
    <span>
      Tab 2
    </span>
  </button>
  <button
    aria-controls="tab-:r0:-panel-3"
    aria-selected="false"
    class="rmd-tab"
    id="tab-:r0:-3"
    role="tab"
    tabindex="-1"
    type="button"
  >
    <span>
      Tab 3
    </span>
  </button>
</div>
`;
